test/host_tests/sal_tcpserver.py

```diff
@@ -0,0 +1,152 @@
+# Copyright 2015 ARM Limited
+#
+# Licensed under the Apache License, Version 2.0
+# See LICENSE file for details.
+
+import sys
+import socket
+import select
+import time
+import threading
+from SocketServer import BaseRequestHandler, TCPServer, _eintr_retry
+from mbed_host_tests import BaseHostTest
+
+
+class TCPServerV4(TCPServer):
+    address_family = socket.AF_INET
+    allow_reuse_address = True
+
+    def __init__(self, server_address, RequestHandlerClass):
+        TCPServer.__init__(self, server_address, RequestHandlerClass)
+        self._shutdown_request = False
+        
+        # watchdog guards against the failure mode when the remote target fails 
+        # to send any packets. If the watchdog reaches the high water mark, the 
+        # server is terminated so as not to leave the server thread unterminated
+        self.watchdog = 0.0
+
+    def serve_forever(self, poll_interval=0.5):
+        """provide an override that can be shutdown from a request handler.
+        The threading code in the BaseSocketServer class prevented this from working
+        even for a non-threaded blocking server.
+        """
+        
+        try:
+            while not self._shutdown_request:
+                r, w, e = _eintr_retry(select.select, [self], [], [], poll_interval)
+                if self in r:
+                    # reset watchdog
+                    self.watchdog = 0.0
+                    self._handle_request_noblock()
+                
+                else:
+                    self.watchdog += poll_interval
+                
+                
+                if self.watchdog > 20.0:
+                    self._shutdown_request = True
+                
+        finally:
+            self._shutdown_request = False
+        
+        
+
+class TCPHandler(BaseRequestHandler):
+    MAX_RX_SIZE = 8192
```

Comment:
- Its appropriate to make rx_len >= client tx_len. 
- in socket_abstract_test.cpp\test\tcp_client_echo\tcp_client_echo.cpp
tx_len_max = SOCKET_SENDBUF_BLOCKSIZE * 2^SOCKET_SENDBUF_ITERATIONS
= 32 * 2^8 = 8192.
- I will document this fact in the code.
- when testing, for tx_len = 8192 there were stack errors. we should fix these before continuing to larger sizes if appropriate.